src/GraphQL.Tests/Types/AutoRegisteringInterfaceGraphTypeTests.cs

```diff
@@ -0,0 +1,809 @@
+    [InlineData("{cat{...frag}} fragment frag on IAnimal { id name }", @"{""data"":{""cat"":{""id"":""10"",""name"":""Fluffy""}}}")]
+    [InlineData("{dog{...frag}} fragment frag on IAnimal { id name }", @"{""data"":{""dog"":{""id"":""20"",""name"":""Shadow""}}}")]
+    public async Task EndToEndTest(string query, string expected)
+    {
+        var services = new ServiceCollection();
+        services.AddGraphQL(b => b
+            .AddAutoSchema<TestQuery>()
+            .AddSystemTextJson());
+        using var provider = services.BuildServiceProvider();
+        var executer = provider.GetRequiredService<IDocumentExecuter>();
+        var result = await executer.ExecuteAsync(new()
+        {
+            Query = query,
+            RequestServices = provider,
+        }).ConfigureAwait(false);
+        var serializer = provider.GetRequiredService<IGraphQLTextSerializer>();
+        var actual = serializer.Serialize(result);
+        actual.ShouldBeCrossPlatJson(expected);
+    }
+
+    [Fact]
+    public async Task ExecutesQueryWithInterfaceOnly()
+    {
+        var services = new ServiceCollection();
+        services.AddGraphQL(b => b
+            .AddAutoSchema<TestQuery2>()
+            .AddSystemTextJson());
+        using var provider = services.BuildServiceProvider();
+        var executer = provider.GetRequiredService<IDocumentExecuter>();
+        var result = await executer.ExecuteAsync(new()
+        {
+            Query = "{find(type:CAT){id name}}",
+            RequestServices = provider,
+        }).ConfigureAwait(false);
+        var serializer = provider.GetRequiredService<IGraphQLTextSerializer>();
+        var actual = serializer.Serialize(result);
+        actual.ShouldBeCrossPlatJson(@"{""data"":{""find"":{""id"":""10"",""name"":""Fluffy""}}}");
+    }
+
+    public class TestQuery
+    {
+        public static IAnimal Find(AnimalType type) => type switch
+        {
+            AnimalType.Cat => Cat(),
+            AnimalType.Dog => Dog(),
+            _ => throw new ArgumentOutOfRangeException(nameof(type)),
+        };
+
+        public static Cat Cat() => new Cat() { Name = "Fluffy", Lives = 9 };
+        public static Dog Dog() => new Dog() { Name = "Shadow", IsLarge = true };
+    }
+
+    public class TestQuery2
+    {
+        public static IAnimal Find(AnimalType type) => type switch
+        {
+            AnimalType.Cat => new Cat() { Name = "Fluffy", Lives = 9 },
+            AnimalType.Dog => new Dog() { Name = "Shadow", IsLarge = true },
+            _ => throw new ArgumentOutOfRangeException(nameof(type)),
+        };
+    }
+
+    public interface IObject
+    {
+        [Id] int Id { get; }
+    }
+
+    public interface IAnimal : IObject
+    {
+        AnimalType Type { get; }
+        string Name { get; }
+    }
+
+    public enum AnimalType
+    {
+        Cat,
+        Dog,
+    }
+
+    [Implements(typeof(IAnimal))]
+    public class Cat : IAnimal
```

Comment:
Notice the explicit interface dependency attribute.  I'm not sure what else to do here; how else to tell what interfaces should be mapped or not?  It would be nice to somehow notice what CLR types have been mapped to GraphQL interfaces and automatically enable those mappings... but that will have to be a future feature.